enclave/core/debugmalloc.c

```diff
@@ -474,4 +486,262 @@ size_t oe_debug_malloc_check(void)
+            _failure_callback(__FILE__, __LINE__, __FUNCTION__, size);
+    }
+
+    return p;
+}
+
+void* oe_memalign(size_t alignment, size_t size)
+{
+    void* ptr = NULL;
+
+    // The only difference between posix_memalign and the obsolete memalign is
+    // that posix_memalign requires alignment to be a multiple of sizeof(void*).
+    // Adjust the alignment if needed.
+    alignment = oe_round_up_to_multiple(alignment, sizeof(void*));
+
+    oe_posix_memalign(&ptr, alignment, size);
+    return ptr;
+}
+
+int oe_posix_memalign(void** memptr, size_t alignment, size_t size)
+{
+    int rc = -1;
+
+    if (oe_use_debug_malloc)
+        rc = oe_debug_posix_memalign(memptr, alignment, size);
+    else
+        rc = oe_posix_memalign(memptr, alignment, size);
+
+    if (rc != 0 && size)
+    {
+        if (_failure_callback)
+            _failure_callback(__FILE__, __LINE__, __FUNCTION__, size);
+    }
+
+    return rc;
+}
+
+size_t oe_malloc_usable_size(void* ptr)
+{
+    if (oe_use_debug_malloc)
+    {
+        return oe_debug_malloc_usable_size(ptr);
+    }
+    else
+    {
+        return oe_allocator_malloc_usable_size(ptr);
+    }
+}
+
+oe_result_t oe_check_memory_leaks(void)
+{
+    if (!oe_disable_debug_malloc_check && oe_debug_malloc_check() != 0)
+        return OE_MEMORY_LEAK;
+    return OE_OK;
+}
+
+oe_result_t oe_debug_malloc_tracking_start(void)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    oe_spin_lock(&_spin);
+    if (!oe_use_debug_malloc_tracking)
+    {
+        oe_use_debug_malloc_tracking = true;
+        result = OE_OK;
+    }
+    oe_spin_unlock(&_spin);
+
+    return result;
+}
+
+oe_result_t oe_debug_malloc_tracking_stop(void)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    oe_spin_lock(&_spin);
+    if (oe_use_debug_malloc_tracking)
+    {
+        oe_use_debug_malloc_tracking = false;
+        result = OE_OK;
+    }
+    oe_spin_unlock(&_spin);
+
+    return result;
+}
+
+static oe_result_t _copy_frames(
+    header_t* p,
+    char** str,
+    size_t* len,
+    size_t* index)
+{
+    oe_result_t result = OE_FAILURE;
+    char** syms = NULL;
+
+    if (!(syms = oe_backtrace_symbols(p->addrs, (int)(p->num_addrs))))
+    {
+        goto done;
+    }
+
+    for (uint64_t i = 0; i < p->num_addrs; i++)
+    {
+        size_t size_s = oe_strlen(syms[i]);
+        size_t size_a = sizeof(p->addrs[i]) * 2;
+        size_t size = size_s + size_a + 6;
+
+        if (*index + size >= *len)
+        {
+            while (*index + size >= *len)
+            {
+                *len *= 2;
+            }
+
+            *str = oe_realloc(*str, *len);
+            if (*str == NULL)
+            {
+                result = OE_ENOMEM;
+                goto done;
+            }
+        }
+
+        oe_snprintf(*str + *index, size, "%s(): %p\n", syms[i], p->addrs[i]);
+        *index = oe_strlen(*str);
+    }
+
+    (*str)[(*index)++] = '\n';
+    (*str)[*index] = '\0';
+
+    result = OE_OK;
+
+done:
+    if (syms)
```

Comment:
should not need line 700, as free apis should always be safe to pass NULL to, like POSIX free() is

Reply:
Fixed, thanks!